### PR TITLE
[PyTorch] Add Float8Tensor option to avoid updating transpose cache when possible

### PR DESCRIPTION
**Pull request by timmoon10**
The NeMo distributed optimizer manually fills the `Float8Tensor` transpose cache at some arbitrary point in the first microbatch:
https://github.com/NVIDIA/NeMo/blob/b100cd1f7c00ba266134b60c98ee7fb8821be5df/nemo/core/optim/distributed_adam.py#L442
This doesn't have much benefit for now, but it does open the door for some future optimizations (improving load balancing with pipeline parallelism, allowing for multi-tensor kernels). However, passing `is_first_microbatch=True` in the backward pass will cause the `Float8Tensor` to recompute the transpose:
https://github.com/NVIDIA/TransformerEngine/blob/a95006170cf36712dcfb4cb27e32431e73125a2c/transformer_engine/pytorch/module/linear.py#L350
~The solution is a `_lazy_transpose_cache` attribute in `Float8Tensor`. If `True`, then calling `transpose` with `update_cache=True` will just return the cached value if it already exists.~ The updated solution is to add more options to the `update_cache` kwarg: `"force"`/`True` (calculate transpose and cache), `"reuse_only"`/`False` (reuse cache if available, otherwise calculate transpose without caching), `"lazy"` (reuse cache if available, otherwise calculate transpose and cache).

There's an argument that `"lazy"` should be the default behavior with `update_cache=True`, but I'm uneasy since an arg name like `update_cache` implies that we update the cache.

### CODE REVIEW

File: transformer_engine/pytorch/float8_tensor.py [hunk 1]
```diff
@@ -478,7 +483,7 @@ def transpose(
             return super().transpose(dim0, dim1)
 
         # Clear cache if needed
-        if update_cache:
+        if update_cache and not self._lazy_transpose_cache:
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/float8_tensor.py:486]: I'm a bit curious, why did we not use a logic like this? If we make sure that we call `_reset_caches` (that makes `_transpose=None`) at the right places (e.g. in-place ops and like how it's called in dist opt in NeMo), shouldn't this be sufficient and we won't need a separate `_lazy_transpose_cache` parameter? 

```suggestion
        if self._transpose is not None:
            return self._transpose
        else:
            out = Float8Tensor.make_like(
                self,
                data=tex.fp8_transpose(
                    self._data.contiguous(),
                    self._fp8_dtype,
                ),
            )
        # Update cache if needed
        if update_cache:
            self._transpose = out
        return out
```

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/float8_tensor.py:486]: Ah I see why, perhaps to support the case when we don't want to cache any transpose

Comment (timmoon10) [reply transformer_engine/pytorch/float8_tensor.py:486]: Mostly I'm trying to stay backward compatible with the existing `update_cache` API and keep its behavior close to its English meaning (it feels wrong if the cache doesn't update when you pass `update_cache=True`).

If I were redoing this, I think I'd replace this with a kwarg called `cache` or `use_cache` that does your proposed behavior. The caching API has ping-ponged a few times though, see https://github.com/NVIDIA/TransformerEngine/commit/00b9c311e2f71c405f49ea8ef19da46fc01515b2.

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/float8_tensor.py:486]: I see the reasoning and I agree that now `update_cache` doesn't seem to be an apt name for this feature. 

I wonder if we could treat `update_cache` as a Union[str, bool] and pass `update_cache="maybe"` but sure it's still substantial design change. 

Another idea could be to have `lazy_transpose_cache` also an argument to `transpose`. But maybe having `_lazy_transpose_cache` an attr of the `Float8Tensor` class is fine too. 

What do you think?

Comment (timmoon10) [reply transformer_engine/pytorch/float8_tensor.py:486]: I like `"maybe"`. It's backwards compatible and we could add other weird configurations in the future as needed (e.g. reset cache without replacing, calculate transpose without replacing cache, reset).

Comment (timmoon10) [reply transformer_engine/pytorch/float8_tensor.py]: I've modified `update_cache` to accept `"force"`, `"reuse_only"`, `"lazy"`. I think the last one is a bit more precise than `"maybe"`.

**Comment by sudhakarsingh27** [comment]
Lgtm!

**Review by sudhakarsingh27 — APPROVED** [review]

**Pull request merged**